<a href="https://colab.research.google.com/github/lj89/CORD19/blob/master/RandPython_CosSimPolymeraseAbstractUniversal_sentence_encode.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import re

In [0]:
# pip install rpy2

In [0]:
# # enables the %%R magic, not necessary if you've already done this
# %load_ext rpy2.ipython

# import pandas as pd
# df = pd.DataFrame({
#     'cups_of_coffee': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
#     'productivity': [2, 5, 6, 8, 9, 8, 0, 1, 0, -1]
# })

In [0]:
# %%R -i df -w 5 -h 5 --units in -r 200
# # import df from global environment
# # make default figure size 5 by 5 inches with 200 dpi resolution

# install.packages("ggplot2", repos='http://cran.us.r-project.org', quiet=TRUE)
# library(ggplot2)
# ggplot(df, aes(x=cups_of_coffee, y=productivity)) + geom_line()

In [0]:
%load_ext rpy2.ipython

The rpy2.ipython extension is already loaded. To reload it, use:
  %reload_ext rpy2.ipython


In [0]:
%%R -i df -w 5 -h 5 --units in -r 200

install.packages("tidyverse", repos='http://cran.us.r-project.org', quiet=TRUE)

install.packages("magrittr", repos='http://cran.us.r-project.org', quiet=TRUE)

install.packages("bigrquery", repos='http://cran.us.r-project.org', quiet=TRUE)

library(tidyverse)
library(magrittr)
library(bigrquery)


In [0]:
# %%R 
# library(tidyverse)


In [0]:
# import tensorflow as tf
# import tensorflow_hub as hub
# import numpy as np
# import tensorflow_text

# questions = ["What is your age?"]
# responses = ["I am 20 years old.", "good morning"]
# response_contexts = ["I will be 21 next year.", "great day."]

# module = hub.load('https://tfhub.dev/google/universal-sentence-encoder-qa/3')

# question_embeddings = module.signatures['question_encoder'](
#             tf.constant(questions))
# response_embeddings = module.signatures['response_encoder'](
#         input=tf.constant(responses),
#         context=tf.constant(response_contexts))

# np.inner(question_embeddings, response_embeddings)


In [0]:
import numpy as np
import pandas as pd
#import the data
# directory = 'C:/Users/N1110/Desktop/CORD19/KeyWordSearchResultsCSV/'
# df = pd.read_csv(directory + 'tibbleABSTRACT_polymerase2391.csv') 

In [0]:
data_path = "/content/drive/My Drive/Colab Notebooks/cord19/data/"
data_filename = data_path +"POLYMERASE_ABSTRACT_TFIDF_10clustersResultsWithOriginalABS.csv"
print(data_filename)


/content/drive/My Drive/Colab Notebooks/cord19/data/POLYMERASE_ABSTRACT_TFIDF_10clustersResultsWithOriginalABS.csv


In [0]:
# allow access to google drive files
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
df = pd.read_csv(data_filename)
df.head(5)

index                                                ABS  ...  G9  G10
0      1  abstract astrocytes produce granulocytemacroph...  ...   0    0
1      2  abstract replication of avian infectious bronc...  ...   0    0
2      3  abstract the infectivity of vesicular stomatit...  ...   0    0
3      4  abstract two temporally and enzymatically dist...  ...   0    0
4      5  abstract rnadependent rna polymerase rdrp acti...  ...   0    0

[5 rows x 42 columns]

In [0]:
df.shape 
# some problem, could not drop duplicates due to same content has different sha number df2=df1.drop_duplicates(keep='last')
# but it seems we drop duplicates in previous steps.

(1941, 42)

In [0]:
df.isnull().sum()

index    0
ABS      0
0        0
1        0
2        0
3        0
4        0
5        0
6        0
7        0
8        0
9        0
10       0
11       0
12       0
13       0
14       0
15       0
16       0
17       0
18       0
19       0
20       0
21       0
22       0
23       0
24       0
25       0
26       0
27       0
28       0
29       0
G1       0
G2       0
G3       0
G4       0
G5       0
G6       0
G7       0
G8       0
G9       0
G10      0
dtype: int64

In [0]:
data1 = np.array(df.ABS.drop_duplicates(keep='last'))
data1

array(['abstract astrocytes produce granulocytemacrophage colonystimulating factor gmcsf and support the survival and proliferation of microglia to study the functions of gmcsf in the central nervous system cns we examined the effects of gmcsf on cytokine production by glial cells gmcsf induced interleukin6 il6 production by microglia but not by astrocytes in a dosedependent manner as assessed by bioassay and the detection of il6 mrna by reverse transcriptasepolymerase chain reaction rtpcr analysis gmcsf did not induce tumor necrosis factor tnfa or il1 in microglia and astrocytes whereas lipopolysaccharide induced all these cytokines the induction of il6 by gmcsf in microglia was completely inhibited by antibodies to gmcsf neither il3 nor macrophagecsf mcsf induced il6 production in microglia given that il1 and tnf monokines derived from microglia induce il6 production in astrocytes but not in microglia results indicate that astrocytes and microglia may mutually regulate il6 production

In [0]:
data1.shape

(1941,)

In [0]:
data1list=data1.tolist()


In [0]:
type(data1list)

list

In [0]:
# data1list.dropna(inplace=True)

# AttributeError: 'list' object has no attribute 'dropna'


In [0]:
data1list

['abstract astrocytes produce granulocytemacrophage colonystimulating factor gmcsf and support the survival and proliferation of microglia to study the functions of gmcsf in the central nervous system cns we examined the effects of gmcsf on cytokine production by glial cells gmcsf induced interleukin6 il6 production by microglia but not by astrocytes in a dosedependent manner as assessed by bioassay and the detection of il6 mrna by reverse transcriptasepolymerase chain reaction rtpcr analysis gmcsf did not induce tumor necrosis factor tnfa or il1 in microglia and astrocytes whereas lipopolysaccharide induced all these cytokines the induction of il6 by gmcsf in microglia was completely inhibited by antibodies to gmcsf neither il3 nor macrophagecsf mcsf induced il6 production in microglia given that il1 and tnf monokines derived from microglia induce il6 production in astrocytes but not in microglia results indicate that astrocytes and microglia may mutually regulate il6 production by di

In [0]:
# datatestlist=['a dosedependent manner as assessed by bioassay and the detection of IL6 mRNA by reverse transcriptasepolymerase chain reaction RTPCR analysis GMCSF did not induce tumor necrosis factor TNFa or IL1 in',
#  'in the presence of amanitin in Chinese hamster ovary cells which possess a drugresistant RNA polymerase II These findings indicate that avian infectious bronchitis virus requires the intact cell nucleus and',
#  'of all five virion proteins particularly the reduced disulfides of the L protein The RNA polymerase activity of intact VSV was inhibited by iodoacetic acid and to a greater extent by']

In [0]:
# datatestlist

In [0]:
import tensorflow as tf
import tensorflow_hub as hub


In [0]:
#Universal Sentence Encoder family

# embed = hub.load("https://tfhub.dev/google/universal-sentence-encoder/3") #this step takes long on local, fast on cloud. worked well


# embed = hub.load("https://tfhub.dev/google/universal-sentence-encoder-large/5") # took some time even on cloud about 1-2 min, got ResourceExhaustedError at next step
#The universal-sentence-encoder-large model is trained with a Transformer encoder. try to run on cloud but got got ResourceExhaustedError# 


#embed = hub.load("https://tfhub.dev/google/universal-sentence-encoder-large/3?tf-hub-format=compressed") #try compressed large, got TypeError: 'AutoTrackable' object is not callable

embed = hub.load("https://tfhub.dev/google/universal-sentence-encoder-qa/3") # ResourceExhaustedError  

# # # ResourceExhaustedError  

In [0]:
embeddings = embed(data1list)["outputs"]

print(embeddings)

ResourceExhaustedError: ignored

In [0]:
embeddings.shape

TensorShape([1941, 512])

In [0]:
NParray1941papers512vector=np.array(embeddings)
NParray1941papers512vector

array([[ 0.02409273,  0.00333242,  0.02617532, ..., -0.05257906,
        -0.02694646, -0.06477401],
       [ 0.00073825,  0.02878782,  0.00838283, ..., -0.05337691,
        -0.03477787,  0.00727509],
       [-0.02900795,  0.03677291,  0.01629622, ..., -0.04934354,
        -0.0058875 , -0.03237623],
       ...,
       [-0.0604236 , -0.0001342 , -0.05867178, ..., -0.05803579,
        -0.06144453, -0.05860412],
       [-0.01348855,  0.04927895, -0.05944376, ..., -0.03317846,
        -0.04884242, -0.04074993],
       [ 0.01909224,  0.0124238 , -0.0579484 , ..., -0.0470332 ,
        -0.00612294, -0.06096034]], dtype=float32)

In [0]:
NParray1941papers512vector[0]

array([ 0.02409273,  0.00333242,  0.02617532, -0.05642438,  0.06698681,
        0.06434409, -0.01563033, -0.0650824 ,  0.06350502,  0.05623564,
        0.06831057, -0.04124111,  0.02260143,  0.02196286, -0.00081022,
       -0.06784237, -0.06504352, -0.04427297,  0.03884536,  0.02333702,
        0.04578838, -0.04443114, -0.05849654, -0.01590162,  0.02861416,
        0.06099916, -0.04841137, -0.00332013, -0.01440635,  0.06032176,
        0.0608208 ,  0.06509712,  0.03341629, -0.0598122 ,  0.02403202,
       -0.01470865, -0.06134612, -0.03351074, -0.03065762, -0.06144067,
        0.04884221,  0.03052049,  0.05471962, -0.06428806, -0.00765103,
        0.0507318 , -0.0644527 , -0.04161062, -0.01791507,  0.04753461,
        0.00709735, -0.00070491, -0.06732775, -0.05811542, -0.02782338,
       -0.06196564, -0.05078926,  0.03742876,  0.04106392, -0.04783525,
        0.03991013,  0.03794602,  0.0526868 , -0.05226299, -0.06610987,
       -0.06487995,  0.00883707,  0.03881599, -0.01578001,  0.02

In [0]:
DF1941papers512vector=pd.DataFrame(np.array(embeddings))
DF1941papers512vector

0         1         2    ...       509       510       511
0     0.024093  0.003332  0.026175  ... -0.052579 -0.026946 -0.064774
1     0.000738  0.028788  0.008383  ... -0.053377 -0.034778  0.007275
2    -0.029008  0.036773  0.016296  ... -0.049344 -0.005888 -0.032376
3     0.025093 -0.005877  0.036179  ... -0.059728 -0.057004 -0.043818
4    -0.020635  0.030479 -0.051886  ... -0.014145 -0.046835 -0.060753
...        ...       ...       ...  ...       ...       ...       ...
1936 -0.044237 -0.020313 -0.048073  ... -0.057656 -0.012462 -0.054203
1937 -0.035588  0.058073 -0.050484  ... -0.047205 -0.041322 -0.023640
1938 -0.060424 -0.000134 -0.058672  ... -0.058036 -0.061445 -0.058604
1939 -0.013489  0.049279 -0.059444  ... -0.033178 -0.048842 -0.040750
1940  0.019092  0.012424 -0.057948  ... -0.047033 -0.006123 -0.060960

[1941 rows x 512 columns]

In [0]:
DF1941papers512vector.to_csv(data_path+"POLYMERASEabstractUni_Sent_Encod_vectors.csv", sep=',',index=False)


In [0]:
testAbstract1=["background enterovirus 71 ev71 is one of the major causative agents of hand foot and mouth disease hfmd which is sometimes associated with severe central nervous system disease in children there is currently no specific medication for ev71 infection quercetin one of the most widely distributed flavonoids in plants has been demonstrated to inhibit various viral infections however investigation of the antiev71 mechanism has not been reported to date methods the antiev71 activity of quercetin was evaluated by phenotype screening determining the cytopathic effect cpe and ev71induced cells apoptosis the effects on ev71 replication were evaluated further by determining virus yield viral rna synthesis and protein expression respectively the mechanism of action against ev71 was determined from the effective stage and timeofaddition assays the possible inhibitory functions of quercetin via viral 2apro 3cpro or 3dpol were tested the interaction between ev71 3cpro and quercetin was predicted and calculated by molecular docking results quercetin inhibited ev71mediated cytopathogenic effects reduced ev71 progeny yields and prevented ev71induced apoptosis with low cytotoxicity investigation of the underlying mechanism of action revealed that quercetin exhibited a preventive effect against ev71 infection and inhibited viral adsorption moreover quercetin mediated its powerful therapeutic effects primarily by blocking the early postattachment stage of viral infection further experiments demonstrated that quercetin potently inhibited the activity of the ev71 protease 3cpro blocking viral replication but not the activity of the protease 2apro or the rna polymerase 3dpol modeling of the molecular binding of the 3cproquercetin complex revealed that quercetin was predicted to insert into the substratebinding pocket of ev71 3cpro blocking substrate recognition and thereby inhibiting ev71 3cpro activity conclusions quercetin can effectively prevent ev71induced cell injury with low toxicity to host cells quercetin may act in more than one way to deter viral infection exhibiting some preventive and a powerful therapeutic effect against ev71 further quercetin potently inhibits ev71 3cpro activity thereby blocking ev71 replication"]

In [0]:
testAbstract2=["background to investigate the effects and immunological mechanisms of the traditional chinese medicine xinjiaxiangruyin on controlling influenza virus fm1 strain infection in mice housed in a hygrothermal environment methods mice were housed in normal and hygrothermal environments and intranasally infected with influenza virus fm1 a highperformance liquid chromatography fingerprint of xinjiaxiangruyin was used to provide an analytical method for quality control realtime quantitative polymerase chain reaction rtqpcr was used to measure messenger rna expression of tolllike receptor 7 tlr7 myeloid differentiation primary response 88 myd88 and nuclear factorkappa b nfb p65 in the tlr7 signaling pathway and virus replication in the lungs western blotting was used to measure the expression levels of tlr7 myd88 and nfb p65 proteins flow cytometry was used to detect the proportion of th17tregulatory cells results xinjiaxiangruyin effectively alleviated lung inflammation in c57bl6 mice in hot and humid environments guizhimahuanggebantang significantly reduced lung inflammation in c57bl6 mice the expression of tlr7 myd88 and nfb p65 mrna in lung tissue of wt mice in the normal environment gzmhgbt group was significantly lower than that in the model group p  005 in wt mice exposed to the hot and humid environment the expression levels of tlr7 myd88 and nfb p65 mrna in the xjxry group were significantly different from those in the virus group the expression levels of tlr7 myd88 and nfb p65 protein in lung tissue of wt mice exposed to the normal environment gzmhgbt group was significantly lower than those in the model group in wt mice exposed to hot and humid environments the expression levels of tlr7 myd88 and nfb p65 protein in xjxry group were significantly different from those in the virus group conclusion guizhimahuanggebantang demonstrated a satisfactory therapeutic effect on mice infected with the influenza a virus fm1 strain in a normal environment and xinjiaxiangruyin demonstrated a clear therapeutic effect in damp and hot environments and may play a protective role against influenza through downregulation of the tlr7 signal pathway"]

In [0]:
Question1=['What are clinical effective therapeutics or drugs for COVID-19?']

In [0]:
embeddingsT1 = embed(testAbstract1)["outputs"]

print(embeddingsT1)

TypeError: ignored

In [0]:
embeddingsT2 = embed(testAbstract2)["outputs"]

print(embeddingsT2)



tf.Tensor(
[[-0.02189232  0.05631374 -0.0377368  -0.03628954  0.04663548  0.00296444
  -0.03610794 -0.05677947  0.04200027  0.03138404  0.05753179 -0.03210878
  -0.05649831  0.00484167 -0.01503811 -0.05743095 -0.05751107 -0.02669043
   0.04074129 -0.02373612 -0.05725251 -0.0556747  -0.04491174 -0.05481869
  -0.00917479  0.00505077  0.00237335  0.05276376 -0.05218472  0.04024547
   0.05591891  0.05753054  0.05584415 -0.05722082 -0.03664611  0.019679
  -0.05545347 -0.00848423 -0.05707751  0.00891382  0.03175837  0.04898379
  -0.0175502   0.02138516 -0.05726926  0.01688037 -0.03017469 -0.01078759
   0.04228764 -0.03835077  0.05693702  0.05441052 -0.05750427  0.03369717
  -0.03082169 -0.04512396 -0.05743494 -0.04210987  0.05721775 -0.04819105
  -0.03465275  0.04978205 -0.00059543  0.05562617  0.05725281 -0.05656446
   0.05694064  0.04700271  0.03840362  0.02071313 -0.01211066  0.05387345
   0.05150827  0.02603205 -0.02391492 -0.0527598   0.04534614 -0.04125691
  -0.05699781  0.0498798   0.

In [0]:
embeddingsQ1 = embed(Question1)["outputs"]

print(embeddingsT2)

tf.Tensor(
[[-0.02189232  0.05631374 -0.0377368  -0.03628954  0.04663548  0.00296444
  -0.03610794 -0.05677947  0.04200027  0.03138404  0.05753179 -0.03210878
  -0.05649831  0.00484167 -0.01503811 -0.05743095 -0.05751107 -0.02669043
   0.04074129 -0.02373612 -0.05725251 -0.0556747  -0.04491174 -0.05481869
  -0.00917479  0.00505077  0.00237335  0.05276376 -0.05218472  0.04024547
   0.05591891  0.05753054  0.05584415 -0.05722082 -0.03664611  0.019679
  -0.05545347 -0.00848423 -0.05707751  0.00891382  0.03175837  0.04898379
  -0.0175502   0.02138516 -0.05726926  0.01688037 -0.03017469 -0.01078759
   0.04228764 -0.03835077  0.05693702  0.05441052 -0.05750427  0.03369717
  -0.03082169 -0.04512396 -0.05743494 -0.04210987  0.05721775 -0.04819105
  -0.03465275  0.04978205 -0.00059543  0.05562617  0.05725281 -0.05656446
   0.05694064  0.04700271  0.03840362  0.02071313 -0.01211066  0.05387345
   0.05150827  0.02603205 -0.02391492 -0.0527598   0.04534614 -0.04125691
  -0.05699781  0.0498798   0.

In [0]:
test1=np.array(embeddingsT1)
test2=np.array(embeddingsT2)

question1=np.array(embeddingsQ1)

In [0]:
# from math import *
# def dot_prod(x,y):
    
#     return np.sum(a*b for a,b in zip(x,y))
# def vec_len(x):
#     return np.sqrt(np.sum([a*a for a in x]))
# def cos_sim(x,y): 
#     return dot_prod(x,y)/float(vec_len(x)*vec_len(y))

In [0]:
result = np.sum(NParray1941papers512vector*test1,axis=1)/(np.sqrt(np.sum(NParray1941papers512vector*NParray1941papers512vector,axis=1))*np.sqrt(np.sum(test1*test1)))

In [0]:
maxRows=result.argsort()[-5:][::-1]  #https://stackoverflow.com/questions/6910641/how-do-i-get-indices-of-n-maximum-values-in-a-numpy-array

In [0]:
maxRows

array([ 984, 1499, 1937, 1064,  663])

In [0]:
result[result.argsort()[-5:][::-1]]

array([1.0000001 , 0.82702947, 0.7755653 , 0.7716375 , 0.7676146 ],
      dtype=float32)

In [0]:
print(df.ABS.iloc[maxRows])

984     background enterovirus 71 ev71 is one of the m...
1499    the ubiquitinproteasome system has been shown ...
1937    therapeutic targeting of host cell factors req...
1064    epidemic and pandemic influenza a virus iav po...
663     background protein ubiquitination andor degrad...
Name: ABS, dtype: object


In [0]:
import textwrap

In [0]:
resultq1 = np.sum(NParray1941papers512vector*question1,axis=1)/(np.sqrt(np.sum(NParray1941papers512vector*NParray1941papers512vector,axis=1))*np.sqrt(np.sum(question1*question1)))
maxRowsq1=resultq1.argsort()[-20:][::-1]  #https://stackoverflow.com/questions/6910641/how-do-i-get-indices-of-n-maximum-values-in-a-numpy-array
print("The indexes for most similar papers are:") 
print(maxRowsq1)
print("\n")
print("The cosine similarity for top 20 papers are:") 
print(resultq1[resultq1.argsort()[-20:][::-1]])
print("\n")
print("For Question:\n")
print(textwrap.fill(Question1[0],100))
print("\nWe found the top 20 most related papers as listed below:\n")
print(df.ABS.iloc[maxRowsq1])

The indexes for most similar papers are:
[ 644  818 1457 1145 1829   66 1774  496 1825  637   84  298 1403 1467
 1263  436 1789  647 1273  217]


The cosine similarity for top 20 papers are:
[0.31485566 0.28105113 0.27972373 0.26491037 0.25267276 0.2485031
 0.24528125 0.23647138 0.23592086 0.23354343 0.2223026  0.22137879
 0.22077084 0.21867059 0.21549027 0.21536084 0.21522383 0.21099536
 0.20463012 0.20306943]


For Question:

What are clinical effective therapeutics or drugs for COVID-19?

We found the top 20 most related papers as listed below:

644     ltpgtltbgtltbgt there is no specific drug that...
818     limited data is available on feline leishmanio...
1457    please cite this paper as wathen et al 2012 an...
1145    background rapid molecular methods have create...
1829    introduction the differentiation of viral from...
66      abstract viruses are major pathogenic agents c...
1774    abstractthe global spread of sarscov2 requires...
496     in this study 26 blood samples 

In [0]:
mostRelatedQ1=df.ABS.iloc[maxRowsq1]

In [0]:
DFmostRelatedQ1=pd.DataFrame(mostRelatedQ1)

In [0]:
DFmostRelatedQ1

ABS
644   ltpgtltbgtltbgt there is no specific drug that...
818   limited data is available on feline leishmanio...
1457  please cite this paper as wathen et al 2012 an...
1145  background rapid molecular methods have create...
1829  introduction the differentiation of viral from...
66    abstract viruses are major pathogenic agents c...
1774  abstractthe global spread of sarscov2 requires...
496   in this study 26 blood samples were collected ...
1825  abstract most viral diseases with the exceptio...
637   background medicines for the treatment of 2019...
84    abstract since the last handbook of clinical n...
298   abstract zika fever a mosquitoborne infectious...
1403  objective to examine whether results of a poly...
1467  in recent months there has been a wealth of pr...
1263  background effective therapeutics for respirat...
436   abstract bovine respirovirus 3 also known as b...
1789  up to 25 of patients with profound neutropenia...
647   lets take inventory on the therapies that are ...
1273  the incidence of hivrelated opportunistic infe...
217   in an avian flu pandemic which drugs could be ...

In [0]:
#DFmostRelatedQ1.to_csv(data_path+"mostRelatedQ1POLYMERASEabstract.csv", sep=',',index=False)

In [0]:
DFmostRelatedQ2=pd.read_csv(data_path+"mostRelatedQ1.csv")

In [0]:
DFmostRelatedQ2

index                                                ABS
0       0  ltpgtltbgtltbgt there is no specific drug that...
1       1  limited data is available on feline leishmanio...
2       2  please cite this paper as wathen et al 2012 an...
3       3  background rapid molecular methods have create...
4       4  introduction the differentiation of viral from...
5       5  abstract viruses are major pathogenic agents c...
6       6  abstractthe global spread of sarscov2 requires...
7       7  in this study 26 blood samples were collected ...
8       8  abstract most viral diseases with the exceptio...
9       9  background medicines for the treatment of 2019...
10     10  abstract since the last handbook of clinical n...
11     11  abstract zika fever a mosquitoborne infectious...
12     12  objective to examine whether results of a poly...
13     13  in recent months there has been a wealth of pr...
14     14  background effective therapeutics for respirat...
15     15  abstract bovine respirovirus 3 also known as b...
16     16  up to 25 of patients with profound neutropenia...
17     17  lets take inventory on the therapies that are ...
18     18  the incidence of hivrelated opportunistic infe...
19     19  in an avian flu pandemic which drugs could be ...

In [0]:
%%R 
extract_text_window <- function(dataframe, keyword, half_window_size) {
  dataframe %>% 
    group_by(index) %>% 
    mutate(WORDS = ABS) %>% 
    separate_rows(WORDS, sep = "[ \n]+") %>% 
    mutate(INDEX = seq(from = 1, to = n(), by = 1.0),
           WINDOW_START = case_when(INDEX - half_window_size < 1 ~ 1,
                                    TRUE ~ INDEX - half_window_size),
           WINDOW_END = case_when(INDEX + half_window_size > max(INDEX) ~ max(INDEX),
                                  TRUE ~ INDEX + half_window_size),
           WINDOW = word(string = ABS, start = WINDOW_START, end = WINDOW_END, sep = "[ \n]+")) %>% 
    ungroup() %>% 
    filter(str_detect(string = WORDS, pattern = regex(keyword, ignore_case = TRUE)))
}

In [0]:
# %%R 
# df2=read.csv("/content/drive/My Drive/Colab Notebooks/cord19/data/mostRelatedQ1POLYMERASEabstract.csv", head = TRUE)

In [0]:
df2=DFmostRelatedQ2

In [0]:
%%R -i df2 -w 5 -h 5 --units in -r 200
#therapeut
tibble=df %>% 
  extract_text_window(keyword = "therapeut", half_window_size = 15)
tibble

#write.csv(tibble,'mostRelatedQ1Window.csv')

# A tibble: 4 x 7
  index ABS               WORDS  INDEX WINDOW_START WINDOW_END WINDOW           
  <int> <chr>             <chr>  <dbl>        <dbl>      <dbl> <chr>            
1     0 ltpgtltbgtltbgt … thera…   204          189        208 inhibiting key t…
2     6 abstractthe glob… thera…    13            1         28 abstractthe glob…
3     6 abstractthe glob… thera…   287          272        293 parp1 inhibitor …
4    14 background effec… thera…     3            1         18 background effec…


In [0]:
result2 = np.sum(NParray1941papers512vector*test2,axis=1)/(np.sqrt(np.sum(NParray1941papers512vector*NParray1941papers512vector,axis=1))*np.sqrt(np.sum(test2*test2)))
maxRows2=result2.argsort()[-10:][::-1]  #https://stackoverflow.com/questions/6910641/how-do-i-get-indices-of-n-maximum-values-in-a-numpy-array
print("The indexes for most similar papers are:") 
print(maxRows2)
print("\n")
print("The cosine similarity for top 5 papers are:") 
print(result2[result2.argsort()[-10:][::-1]])
print("\n")
print("For Paper Abstract:\n")
print(textwrap.fill(testAbstract2[0],100))
print("\nWe found the top 10 most similar papers as listed below:\n")
print(df.ABS.iloc[maxRows2])

The indexes for most similar papers are:
[1046 1453 1279 1383 1064 1168 1563  400  210 1419]


The cosine similarity for top 5 papers are:
[1.         0.7110584  0.68514544 0.6791583  0.67234355 0.6699095
 0.6660905  0.6534609  0.64800745 0.6457798 ]


For Paper Abstract:

background to investigate the effects and immunological mechanisms of the traditional chinese
medicine xinjiaxiangruyin on controlling influenza virus fm1 strain infection in mice housed in a
hygrothermal environment methods mice were housed in normal and hygrothermal environments and
intranasally infected with influenza virus fm1 a highperformance liquid chromatography fingerprint
of xinjiaxiangruyin was used to provide an analytical method for quality control realtime
quantitative polymerase chain reaction rtqpcr was used to measure messenger rna expression of
tolllike receptor 7 tlr7 myeloid differentiation primary response 88 myd88 and nuclear factorkappa b
nfb p65 in the tlr7 signaling pathway and virus replicat

In [0]:
result1 = np.sum(NParray1941papers512vector*test1,axis=1)/(np.sqrt(np.sum(NParray1941papers512vector*NParray1941papers512vector,axis=1))*np.sqrt(np.sum(test1*test1)))
maxRows1=result1.argsort()[-10:][::-1]  #https://stackoverflow.com/questions/6910641/how-do-i-get-indices-of-n-maximum-values-in-a-numpy-array
print("The indexes for most similar papers are:") 
print(maxRows1)
print("\n")
print("The cosine similarity for top 5 papers are:") 
print(result1[result1.argsort()[-10:][::-1]])
print("\n")
print("For Paper Abstract:\n")
print(textwrap.fill(testAbstract1[0],100))
print("\nWe found the top 10 most similar papers as listed below:\n")
print(df.ABS.iloc[maxRows1])

The indexes for most similar papers are:
[ 984 1499 1937 1064  663 1636  154   38  146  776]


The cosine similarity for top 5 papers are:
[1.0000001  0.82702947 0.7755653  0.7716375  0.7676146  0.7673607
 0.75756603 0.7570241  0.75692147 0.7561372 ]


For Paper Abstract:

background enterovirus 71 ev71 is one of the major causative agents of hand foot and mouth disease
hfmd which is sometimes associated with severe central nervous system disease in children there is
currently no specific medication for ev71 infection quercetin one of the most widely distributed
flavonoids in plants has been demonstrated to inhibit various viral infections however investigation
of the antiev71 mechanism has not been reported to date methods the antiev71 activity of quercetin
was evaluated by phenotype screening determining the cytopathic effect cpe and ev71induced cells
apoptosis the effects on ev71 replication were evaluated further by determining virus yield viral
rna synthesis and protein expression

# universal-sentence-encoder results
embed = hub.load("https://tfhub.dev/google/universal-sentence-encoder/3")

The indexes for most similar papers are:
[ 984 1499 1937 1064  663 1636  154   38  146  776]


The cosine similarity for top 5 papers are:
[1.0000001  0.8270295  0.7755654  0.7716375  0.76761466 0.7673608
 0.75756603 0.7570242  0.7569215  0.7561373 ]


For Paper Abstract:

background enterovirus 71 ev71 is one of the major causative agents of hand foot and mouth disease
hfmd which is sometimes associated with severe central nervous system disease in children there is
currently no specific medication for ev71 infection quercetin one of the most widely distributed
flavonoids in plants has been demonstrated to inhibit various viral infections however investigation
of the antiev71 mechanism has not been reported to date methods the antiev71 activity of quercetin
was evaluated by phenotype screening determining the cytopathic effect cpe and ev71induced cells
apoptosis the effects on ev71 replication were evaluated further by determining virus yield viral
rna synthesis and protein expression respectively the mechanism of action against ev71 was
determined from the effective stage and timeofaddition assays the possible inhibitory functions of
quercetin via viral 2apro 3cpro or 3dpol were tested the interaction between ev71 3cpro and
quercetin was predicted and calculated by molecular docking results quercetin inhibited ev71mediated
cytopathogenic effects reduced ev71 progeny yields and prevented ev71induced apoptosis with low
cytotoxicity investigation of the underlying mechanism of action revealed that quercetin exhibited a
preventive effect against ev71 infection and inhibited viral adsorption moreover quercetin mediated
its powerful therapeutic effects primarily by blocking the early postattachment stage of viral
infection further experiments demonstrated that quercetin potently inhibited the activity of the
ev71 protease 3cpro blocking viral replication but not the activity of the protease 2apro or the rna
polymerase 3dpol modeling of the molecular binding of the 3cproquercetin complex revealed that
quercetin was predicted to insert into the substratebinding pocket of ev71 3cpro blocking substrate
recognition and thereby inhibiting ev71 3cpro activity conclusions quercetin can effectively prevent
ev71induced cell injury with low toxicity to host cells quercetin may act in more than one way to
deter viral infection exhibiting some preventive and a powerful therapeutic effect against ev71
further quercetin potently inhibits ev71 3cpro activity thereby blocking ev71 replication

We found the top 10 most similar papers as listed below:

984     background enterovirus 71 ev71 is one of the m...
1499    the ubiquitinproteasome system has been shown ...
1937    therapeutic targeting of host cell factors req...
1064    epidemic and pandemic influenza a virus iav po...
663     background protein ubiquitination andor degrad...
1636    avian coronavirus infectious bronchitis virus ...
154     abstract murine coronavirus mouse hepatitis vi...
38      abstract papaverine an inhibitor of camp phosp...
146     abstract footandmouth disease virus fmdv is th...
776     background public health risks associated to i...



The indexes for most similar papers are:
[1046 1453 1279 1383 1064 1168 1563  400  210 1419]


The cosine similarity for top 5 papers are:
[1.         0.7110583  0.6851454  0.6791583  0.6723436  0.6699095
 0.6660905  0.6534609  0.64800745 0.6457798 ]


For Paper Abstract:

background to investigate the effects and immunological mechanisms of the traditional chinese
medicine xinjiaxiangruyin on controlling influenza virus fm1 strain infection in mice housed in a
hygrothermal environment methods mice were housed in normal and hygrothermal environments and
intranasally infected with influenza virus fm1 a highperformance liquid chromatography fingerprint
of xinjiaxiangruyin was used to provide an analytical method for quality control realtime
quantitative polymerase chain reaction rtqpcr was used to measure messenger rna expression of
tolllike receptor 7 tlr7 myeloid differentiation primary response 88 myd88 and nuclear factorkappa b
nfb p65 in the tlr7 signaling pathway and virus replication in the lungs western blotting was used
to measure the expression levels of tlr7 myd88 and nfb p65 proteins flow cytometry was used to
detect the proportion of th17tregulatory cells results xinjiaxiangruyin effectively alleviated lung
inflammation in c57bl6 mice in hot and humid environments guizhimahuanggebantang significantly
reduced lung inflammation in c57bl6 mice the expression of tlr7 myd88 and nfb p65 mrna in lung
tissue of wt mice in the normal environment gzmhgbt group was significantly lower than that in the
model group p  005 in wt mice exposed to the hot and humid environment the expression levels of tlr7
myd88 and nfb p65 mrna in the xjxry group were significantly different from those in the virus group
the expression levels of tlr7 myd88 and nfb p65 protein in lung tissue of wt mice exposed to the
normal environment gzmhgbt group was significantly lower than those in the model group in wt mice
exposed to hot and humid environments the expression levels of tlr7 myd88 and nfb p65 protein in
xjxry group were significantly different from those in the virus group conclusion
guizhimahuanggebantang demonstrated a satisfactory therapeutic effect on mice infected with the
influenza a virus fm1 strain in a normal environment and xinjiaxiangruyin demonstrated a clear
therapeutic effect in damp and hot environments and may play a protective role against influenza
through downregulation of the tlr7 signal pathway

We found the top 10 most similar papers as listed below:

1046    background to investigate the effects and immu...
1453    highly pathogenic h5n1 influenza a virus remai...
1279    morbidity and mortality resulting from influen...
1383    influenza virus infection causes global inhibi...
1064    epidemic and pandemic influenza a virus iav po...
1168    introduction ribavirin rbv is a broadspectrum ...
1563    borna disease virus bdv can persistently infec...
400     abstract background severe acute respiratory s...
210     summary objective human infections caused by a...
1419    rationale respiratory syncytial virus rsv and ...
Name: ABS, dtype: object

In [0]:
# # cos_sim(embeddingsT1,DF1941papers512vector.iloc[0])

# cos_sim(test1,test2)

In [0]:
# result = []
# for ii in range(1941):
#   result.append(cos_sim(test1,DF1941papers512vector.iloc[0]))
# print(result)

In [0]:
# type(test1)

In [0]:
# type(NParray1941papers512vector[0])

In [0]:
# test1.shape

In [0]:
# NParray1941papers512vector[0].shape

In [0]:
# np.reshape(NParray1941papers512vector[0],[1,512]).shape

In [0]:
# DF1941papers512vector.iloc[0].shape

In [0]:
# np.reshape(DF1941papers512vector.iloc[0],[1,512]).shape

In [0]:
# embeddingsT1.shape